src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/EnumTypeModelBinder.cs

```diff
@@ -0,0 +1,121 @@
+        {
+            if (bindingContext == null)
+            {
+                throw new ArgumentNullException(nameof(bindingContext));
+            }
+
+            var valueProviderResult = bindingContext.ValueProvider.GetValue(bindingContext.ModelName);
+            if (valueProviderResult == ValueProviderResult.None)
+            {
+                // no entry
+                return Task.CompletedTask;
+            }
+
+            bindingContext.ModelState.SetModelValue(bindingContext.ModelName, valueProviderResult);
+            var modelType = bindingContext.ModelMetadata.UnderlyingOrModelType;
+            var stringValue = valueProviderResult.FirstValue;
+
+            try
+            {
+                object model;
+                if (string.IsNullOrWhiteSpace(stringValue))
+                {
+                    model = null;
+                }
+                else
+                {
+                    var typeConverter = TypeDescriptor.GetConverter(modelType);
+                    model = typeConverter.ConvertFrom(
+                        context: null,
+                        culture: valueProviderResult.Culture,
+                        value: stringValue);
+                }
+
+                // Check if the converted value is indeed defined on the enum as EnumConverter converts 
+                // value to the backing type (ex: integer) and does not check if the value is defined on the enum.
+                if (!_mvcOptions.AllowBindingUndefinedValueToEnumType && model != null)
+                {
+                    var isFlagsEnum = modelType.IsDefined(typeof(FlagsAttribute), inherit: false);
+                    if (isFlagsEnum)
+                    {
+                        // From EnumDataTypeAttribute.cs in CoreFX
+                        var underlying = Convert.ChangeType(
+                            model,
+                            Enum.GetUnderlyingType(modelType),
+                            valueProviderResult.Culture).ToString();
+                        var converted = model.ToString();
+                        if (string.Equals(underlying, converted, StringComparison.OrdinalIgnoreCase))
+                        {
+                            model = null;
+                        }
+                    }
+                    else
+                    {
+                        if (!Enum.IsDefined(modelType, model))
+                        {
+                            model = null;
+                        }
+                    }
+                }
+
+                // When converting newModel a null value may indicate a failed conversion for an otherwise required
+                // model (can't set a ValueType to null). This detects if a null model value is acceptable given the
+                // current bindingContext. If not, an error is logged.
+                if (model == null && !bindingContext.ModelMetadata.IsReferenceOrNullableType)
```

Comment:
Shouldn't we still have an error if the value is provided for a nullable enum - but the conversion failed?